streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowReduce.java

```diff
@@ -87,22 +90,33 @@ public void process(final K key, final V value) {
 
             // first get the matching windows
             final long timestamp = context().timestamp();
+            final long expiryTime = internalProcessorContext.streamTime() - windows.maintainMs();
+
             final Map<Long, W> matchedWindows = windows.windowsFor(timestamp);
 
             // try update the window, and create the new window for the rest of unmatched window that do not exist yet
             for (final Map.Entry<Long, W> entry : matchedWindows.entrySet()) {
-                final V oldAgg = windowStore.fetch(key, entry.getKey());
-
-                final V newAgg;
-                if (oldAgg == null) {
-                    newAgg = value;
+                final Long windowStart = entry.getKey();
+                if (windowStart > expiryTime) {
```

Comment:
This has probably always been the case, but the code here and the code above in `KStreamWindowAggregate` are practically the same. Can we introduce an interface and generalize it such that we only do this check etc in one place?

Reply:
Looking more closely at it, I agree. Reduction is a special case of aggregation, so maybe I'll just make this class subclass `KStreamWindowAggregate`, filling in a null initializer and wrapping the reducer. We could actually eliminate the `KStreamWindowAggregateReducer` class entirely, but that would change the logger class and thus require a KIP.